src/ast/passes/codegen_llvm.cpp

```diff
@@ -4197,9 +4211,29 @@ void CodegenLLVM::probereadDatastructElem(Value *src_data,
                                           location loc,
                                           const std::string &temp_name)
 {
-  Value *src = b_.CreateAdd(src_data, offset);
+  // FIXME(amscanne): Pointers are often casted into integer types, and
+  // therefore we need to handle this case. Unfortunately, this means that we
+  // need to rewrap in the intrinsic to ensure that these continue to preserve
+  // offsets during access.  The solution to this to have have the local types
+  // accurately reflect the true types, rather than using `codegen_types`, and
+  // simply apply appropriate conversions at the moment needed.
```

Comment:
Not sure if I understand what this means. What is a 'codegen_type' ?

Reply:
Ah, the `emit_codegen_types` parameter for `GetType`, which coerces pointer types into `Int64` by default (https://github.com/bpftrace/bpftrace/blob/master/src/ast/irbuilderbpf.cpp#L335).

I dropped the code that coerces the context immediately into a `Int64` as a builtin (and all tests are passing, so hopefully this is fine), but unfortunately that integer cast is still going to happen with any kind of variable assignment, e.g.

```
BEGIN {
  @test = 1;
}

tracepoint:syscalls:sys_enter_kill {
  $a = args;
  print($a.pid);
  exit();
}
```

Therefore it necessary to coerce back to a pointer and rewrap in the intrinsic.

Reply:
The question is whether we should completely drop this "store pointers as integers" thing. I'm not sure if has any benefits at all. I know it's probably out of the scope of this PR but it would allow to simplify things here. I'm just wondering how much work it would be to convert everything to use raw pointers and not ints.

Reply:
I'm strongly in favor of that, but I think that's a bigger exercise (and would probably be much easier with some broader cleanup first, to make it easier to reason about the types). I've refactored this change to try to reduce the scope a bit, as I'm looking for a relatively narrow fix. (Happy to push separately on the broader bits.)